solutions/usaco-599.mdx

```diff
@@ -31,127 +31,172 @@ a one-base polynomial hash to check if the substring appears only once. If we
 find an appropriate substring, we find the total distance that we travelled
 before finding out where we were.
 
+## Implementation
 **Time Complexity:** $\mathcal{O}(N^3)$
 
+<LanguageSection>
+<CPPSection>
+
 ```cpp
 #include <bits/stdc++.h>
-
 using namespace std;
 
-typedef pair<int, int> pii;
-#define f first
-#define s second
 using ll = long long;
 
-const ll X = 97;
-const ll M = 999999929;
+const ll P = 31;
+const ll MOD = 1e9 + 7;
+const ll MX = 1e5 * 2;
 
-int n, m, H[505], p[505];
-vector<int> k;
-pii arr[205];
+struct coord {
+	int x, y;
+};
 
-void gen_hash()
-{
-	p[0] = 1;
-	for(int i = 1; i < m; i++)
-	{
-		p[i] = int(p[i-1]*X%M);
-	}
-	H[0] = k[0];
-	for(int i = 1; i < m; i++)
-	{
-		H[i] = int((H[i-1]*X+k[i]) % M);
+signed main() {
+	freopen("lightsout.in", "r", stdin);
+	freopen("lightsout.out", "w", stdout);
+
+	cin.tie(0)->sync_with_stdio(0);
+
+	int n;
+	cin >> n;
+
+	vector<coord> dt(n);
+	for (int i = 0; i < n; i++) {
+		cin >> dt[i].x >> dt[i].y;
 	}
-}
 
-int cwdist[205][205], dist[205];
-void gen_dist()
-{
-	for(int i = 0; i < n; i++)
-	{
-		for(int j = i; j < n; j++)
-		{
-			cwdist[i][(j+1)%n] = cwdist[i][j] + abs((arr[j].f - arr[(j+1)%n].f) + (arr[j].s - arr[(j+1)%n].s));
-			if(j+1 != n) cwdist[(j+1)%n][i] = INT_MAX;
+	function<int(int i, int j)> calc_dist = [&](int i, int j) {
+		return abs(dt[i].x - dt[j].x) + abs(dt[i].y - dt[j].y);
+	};
+
+	vector<vector<vector<int>>> dst(n,
+																	vector<vector<int>>(n,
+																											vector<int>(2, INT32_MAX)));
+
+	// i, j, 0 - clockwise dist from i -> j
+	// i, j, 1 - counter-clockwise dist from i -> j
```

Comment:
are these array indices or smth?

Reply:
yeaaaaah

Reply:
can you change it so that's more explicit